man/man1/mbdatalist.1
.TH mbdatalist 1 "26 January 2010" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbdatalist\fP - parses recursive datalist files and outputs the
complete list of data files, formats, and file weights.

.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbdatalist\fP [\fB-C\fP \fB-F\fP\fIformat\fP \fB-I\fP\fIfilename\fP \fB-N -O -P -Q -R -S -U -Y -Z -V -H\fP]

.SH DESCRIPTION
\fBMBdatalist\fP is a utility for parsing datalist files. 
Datalist files, or lists of swath data files and their format ids, 
are used by a number
of \fBMB-System\fP programs. These lists may contain references
to other datalists, making them recursive. See the \fBMB-System\fP
manual page for details on the format and structure of datalists. 
The program \fBmbdatalist\fP outputs each 
swath data filename, format id, and file
weight encountered as it descends through the input datalist tree. If a
swath data file rather than a datalist is provided as input, the
same swath data filename and format will be the sole output.

This program
can be used in shellscripts to read datalists in the same fashion
as \fBMB-System\fP programs like \fBmbgrid\fP and \fBmbprocess\fP.
This program can also be used to check and debug complex recursive
datalist structures.

The program \fBmbprocess\fP operates on "raw" swath data files,
producing a "processed" swath data file (see the \fBmbprocess\fP man 
page for explanation).
The \fBMB-System\fP algorithm for reading datalists will, 
if a flag is set, replace a swath file name with the associated
"processed" file name when that "processed" file exists. This
flag may be set by embedding "$PROCESSED" as a line in a datalist
or it may be set first by the calling program. The flag may also
be set to preclude reporting "processed" file names (embedding "$RAW"
in a datalist accomplishes this). When setting this flag within
datalists, the first encounter of a $PROCESSED or $RAW tag will
prevail over later instances of either tag. The \fB-P\fP and \fB-U\fP 
options force \fBmbdatalist\fP to output processed file names 
when they exist (\fB-P\fP) or to only output unprocessed 
(raw) file names (\fB-U\fP).

Programs such as \fBmbgrid\fP try to check statistics or "inf" 
files to see if the corresponding data files include data within
the specified geographic bounds. Other programs look for
"fast bathymetry" or "fast navigation" ("fbt" or "fnv") files
in order to read the data more quickly. The \fB-N\fP option causes 
\fBmbdatalist\fP to create these three types of ancillary files 
for each swath data file. The \fB-O\fP option causes \fBmbdatalist\fP 
to create the "inf", "fbt", and "fnv" files only when they don't 
already exist or are out of date (older than the data file).

Datalists may also contain a third value, called the grid
weight, which is used by \fBmbgrid\fP to priortize data.
The larger the grid weight, the more importance \fBmbgrid\fP
attaches to the related bathymetry data. Grid weights can
be applied to datalist entries which are themselves datalist
files, causing these weights to be associated with all of
files referenced therein. However, the default behavior is
for any grid weight in a particular datalist entry to override
values derived from higher levels in the recursive structure.
This behavior can be reversed if a $NOLOCALWEIGHT tag is placed
in the datalist, or in a datalist higher up in the structure.
See the \fBMB-System\fP manual page for a more complete
description.

The \fB-R\fP\fIw/e/s/n\fP option causes the program to check each
data file with an "inf" file for overlap with the desired bounds, 
and only report those files with data in the desired area (or no 
"inf" file to check). This behavior mimics that of \fBmbgrid\fP, 
allowing users to check what data files will contribute to gridding 
some particular area.

The \fB-Q\fP option causes the program to check each data file for 
the existence of any ancillary files (e.g. navigation files, edit 
save files, etc.) referenced in its mbprocess parameter file 
(if the parameter file exists). The program will list any problem
found with the processing parameters, and will also list any
data problem noted in the "inf" files. The possible data problems
include:
        No survey data found
        Zero longitude or latitude in survey data
        Instantaneous speed exceeds 25 km/hr
        Average speed exceeds 25 km/hr
        Sounding depth exceeds 11000 m
        Unsupported Simrad datagram

The \fB-Z\fP option causes the program to generate a datalist file
named "datalistp.mb-1" and then exit. This datalist has the following form:
.br
 	$PROCESSED
 	datalist.mb-1 -1
.br
This file is a commonly used convenience because it allows users to easily
reference the swath files listed (directly or recursively) through
the datalist "datalist.mb-1" with the $PROCESSED flag on. So, in order to
grid the processed bathymetry rather than the raw bathymetry, run \fBmbgrid\fP
with "datalistp.mb-1" as the input rather than "datalist.mb-1".

The \fB-S\fP option causes \fBmbdatalist\fP to report the status of the files
it lists, including whether the file is up to date or needs reprocessing,
and if the file is locked. \fBMBprocess\fP sets locks while operating on 
a swath file to prevent other instances of \fBmbprocess\fP from simultaneously
operating on that same file. This allows one to run \fmbprocess\fP multiple
times simultaneously on a single datalist, either on a single multiprocessor
machine or on multiple computers mounting the same filesystem. 
The consists of creating a small text file
named by appending ".lck" to the swath filename; while this file exists
other programs will not modify the locked file. The locking program deletes
the lock file when it is done. Orphaned lock files may be left if \fBmbprocess\fP
crashes or is interrupted. These will prevent reprocessing by \fBmbprocess\fP, 
but can be both detected with the \fB-S\fP option and removed using the \fB-Y\fP option.

Finally, this program can be used to copy the swath files referenced in a
datalist structure to a single directory and to create a datalist there
(names "datalist.mb-1") that references those swath files. This is 
accomplished using the \fI-C\fP option. The \fI-C\fP copy function will
not be done if the \fI-N\fP, \fI-O\fP, or \fI-Q\fP options are specified,
but is compatible with the \fI-P\fP, \fI-R\fP, and \fI-U\fP options.

.SH AUTHORSHIP
David W. Caress (caress@mbari.org)
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes (dale@ldeo.columbia.edu)
.br
  Lamont-Doherty Earth Observatory

.SH OPTIONS
.TP
.B \-C
.br
Causes the swath files referenced in the input datalist structure to be 
copied to the current directory and creates a datalist
(names "datalist.mb-1") that references the copied swath files. 
The copy function will not be done if the \fI-N\fP, \fI-O\fP, or \fI-Q\fP options are specified.
If the \fI-P\fP, \fI-R\fP, and \fI-U\fP options are specified these
functions will modify which swath files are copied. Any ancilliary files
(e.g. *inf metadata files) will also be copied, but processed data
files derived from the target copied files will not be copied.
.TP
.B \-F
\fIformat\fP
.br
Sets the data format associated with the datalist or swath data
file specified with the \fB-I\fP option. By default, this program
will attempt to determine the format from the input file suffix
(e.g. a file ending in .mb57 has a format id of 57, and a file 
ending in .mb-1 has a format id of -1). A datalist has a format id
of -1. 
.TP
.B \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-I
\fIfilename\fP
.br
Sets the input filename. If \fIformat\fP > 0 (set with the 
\fB-f\fP option) then the swath data filename specified by \fIinfile\fP 
is output along with its format and a file weight of 1.0. 
If \fIformat\fP < 0, then \fIinfile\fP
is treated as a datalist file containing a list of the input swath sonar
data files to be processed and their formats.  The program will parse 
the datalist (recursively, if necessary) and output each swath filename
and the associated format and file weight. 
.TP
.B \-N
This argument causes \fBMBdatalist\fP to generate three types of ancillary 
data files ("inf", "fbt", and "fnv").
In all cases,
the ancillary filenames are just the original filename with 
".inf", ".fbt", or ".fnv" appended on the end.
\fBMB-System\fP makes use of ancillary data files in a number 
of instances. The most prominent ancillary files are metadata or
"inf" files (created from the output of \fBmbinfo\fP). 
Programs such as \fBmbgrid\fP and \fBmbm_plot\fP try to check "inf" 
files to see if the corresponding data files include data within
desired areas. Additional ancillary files are used to speed 
plotting and gridding functions. The "fast bath" or "fbt" files
are generated by copying the swath bathymetry to a sparse,
quickly read format (format 71). The "fast nav" or "fnv" files
are just ASCII lists of navigation generated using \fBmblist\fP
with a \fB-O\fP\fItMXYHSc\fP option. Programs such as \fBmbgrid\fP,
\fBmbswath\fP, and \fBmbcontour\fP will try to read "fbt" and "fnv" files
instead of the full data files whenever only bathymetry or
navigation information are required.
.TP
.B \-O
This argument causes \fBMBdatalist\fP to generate the three ancillary
data files ("inf", "fbt", and "fnv") if
these files don't already exist or are out of date. 
.TP
.B \-P
Normally, \fBmbdatalist\fP allows $PROCESSED and $RAW tags within
the datalist files to determine whether processed file names are
reported when available ($PROCESSED) or only raw file names are
reported ($RAW). The \fB-P\fP option forces \fBmbdatalist\fP 
to output processed file names when they exist.
.TP
.B \-Q
This option causes the program to check each
data file for the existence of any ancillary files 
referenced in its mbprocess parameter file 
(if the parameter file exists). The relevant ancillary
files include edit save files generated by \fBmbedit\fP
or \fBmbclean\fP, navigation files generated by \fBmbnavedit\fP
or \fBmbnavadjust\fP, tide files, and svp files. An error message
is output for each missing ancillary file.
.TP
.B \-R
\fIw/e/s/n\fP
.br
The bounds of the desired area are set in longitude
and latitude using w=west, e=east, s=south, and n=north.
This option causes the program to check each
data file with an "inf" file for overlap with 
the desired bounds, and only report those
files with data in the desired area (or no 
"inf" file to check). This behavior mimics that of
\fBmbgrid\fP, allowing users to check what data files will 
contribute to gridding some particular area.
.TP
.B \-S
This option causes \fBmbdatalist\fP to report the status of the files
it lists, including whether the file is up to date or needs reprocessing,
and if the file is locked. \fBMBprocess\fP sets locks while operating on 
a swath file to prevent other instances of \fBmbprocess\fP from simultaneously
operating on that same file. Locking consists of creating a small text file
named by appending ".lck" to the swath filename; while this file exists
other programs will not modify the locked file. The locking program deletes
the lock file when it is done. Orphaned lock files may be left if \fBmbprocess\fP
crashes or is interrupted. These will prevent reprocessing by \fBmbprocess\fP, 
but can be both detected and removed using \fBmbdatalist\fP.
.TP
.B \-U
Normally, \fBmbdatalist\fP allows $PROCESSED and $RAW tags within
the datalist files to determine whether processed file names are
reported when available ($PROCESSED) or 
only (raw) unprocessed file names are
reported ($RAW). The \fB-U\fP option forces \fBmbdatalist\fP 
to only output raw file names.
.TP
.B \-V
Normally, \fBmbdatalist\fP only prints out the filenames and formats.  
If the \fB-V\fP flag is given, then \fBmbinfo\fP works in a "verbose" mode and
outputs the program version being used.
.TP
.B \-Y
This option causes \fBmbdatalist\fP to remove any processing locks on files
it parses. \fBMBprocess\fP and other programs may set locks while
operating on a swath file to prevent other programs from simultaneously
operating on that same file.The consists of creating a small text file
named by appending ".lck" to the swath filename; while this file exists
other programs will not modify the locked file. The locking program deletes
the lock file when it is done. Orphaned lock files may be left if \fBMB-System\fP
programs crash or are interrupted. These can be detected using the \fB-S\fP
option of \fBmbdatalist\fP.
.TP
.B \-Z
The \fB-Z\fP option causes the program to generate a datalist file that
will first set a $PROCESSED flag and then reference the input
file specified using the \fB-I\fP\fIfilename\fP option. 
The output datalist is named 
by adding a "p.mb-1" suffix to the root of the input file (the root is
the portion before any \fBMB-System\fP suffix).
.br
By default, the input is assumed to be a datalist named datalist.mb-1, resulting
in an output datalist named datalistp.mb-1 with the following contents:
.br
 	$PROCESSED
 	datalist.mb-1 -1
.br
If the input file is specified as a datalist like 
datalist_sslo.mb-1, then the output datalist 
datalist_sslop.mb-1
will have the following contents:
.br
 	$PROCESSED
 	datalist_sslo.mb-1 -1
.br
If the input file is specified as a swath file like 
20050916122920.mb57, then the output datalist 20050916122920p.mb-1
will have the following contents:
.br
 	$PROCESSED
 	20050916122920.mb57 57

.SH EXAMPLES
Suppose we have two swath data files from an EM3000 multibeam
and another two from an Hydrosweep MD multibeam. We might
construct two datalist files. For the EM3000 we might
have a file dlst_em3000.mb-1 containing:
 	0004_20010705_165004_raw.mb57 57
 	0005_20010705_172010_raw.mb57 57
.br
For the Hydrosweep MD data we might have a file dlst_hsmd.mb-1
containing:
 	al10107051649.mb102 102
 	al10107051719.mb102 102

Further suppose that we have found it necessary to edit
the bathymetry in 0005_20010705_172010_raw.mb57 and
al10107051719.mb102 using \fBmbedit\fP, and that \fBmbprocess\fP
has been run on both files to generate processed files
called 0005_20010705_172010_rawp.mb57 and
al10107051719p.mb102. 

If we run:
 	mbdatalist -I dlst_em3000.mb-1
.br
the output is:
 	0004_20010705_165004_raw.mb57 57 1.000000
 	0005_20010705_172010_raw.mb57 57 1.000000
.br
Here the file name is followed by the format and then
by a third column containing the default file weight of 1.0.

Similarly, if we run:
 	mbdatalist -I dlst_hsmd.mb-1
.br
the output is:
 	al10107051649.mb102 102 1.000000
 	al10107051719.mb102 102 1.000000

If we insert a line
 	$PROCESSED
.br
at the top of both dlst_hsmd.mb-1 and dlst_em3000.mb-1,
then the output of \fBmbdatalist\fP changes so that:
 	mbdatalist -I dlst_em3000.mb-1
.br
yields:
 	0004_20010705_165004_raw.mb57 57 1.000000
 	0005_20010705_172010_rawp.mb57 57 1.000000
and:
 	mbdatalist -I dlst_hsmd.mb-1
.br
yields:
 	al10107051649.mb102 102 1.000000
 	al10107051719p.mb102 102 1.000000

Now suppose we create a datalist file called dlst_all.mb-1 
that refers to the two datalists shown above (without the
$PROCESSED tags). If the contents of dlst_all.mb-1 are:
 	dlst_em3000.mb-1 -1 100.0
 	dlst_hsmd.mb-1   -1   1.0
.br
where we have specified different file weights for the
two datalists, then:
 	mbdatalist -I dlst_all.mb-1
.br
yields:
 	0004_20010705_165004_raw.mb57 57 100.000000
 	0005_20010705_172010_raw.mb57 57 100.000000
 	al10107051649.mb102 102 1.000000
 	al10107051719.mb102 102 1.000000

Now, if we use the \fB-P\fP option to force \fBmbdatalist\fP
to output processed data file names when possible, then:
 	mbdatalist -I dlst_all.mb-1 -P
.br
yields:
 	0004_20010705_165004_raw.mb57 57 100.000000
 	0005_20010705_172010_rawp.mb57 57 100.000000
 	al10107051649.mb102 102 1.000000
 	al10107051719p.mb102 102 1.000000

.SH SEE ALSO
\fBmbsystem\fP(l)

.SH BUGS
No true bugs here, only distantly related arthropods... Yum.
Seriously, it would be better if the copy function preserved
the modification times of the copied swath files and ancilliary
files. Copying of processed files should also be an option.


src/utilities/mbdatalist.c
/*--------------------------------------------------------------------
 *    The MB-system:	mbdatalist.c	10/10/2001
 *    $Id: mbdatalist.c,v 5.11 2007/07/05 19:16:19 caress Exp $
 *
 *    Copyright (c) 2001-2009 by
 *    David W. Caress (caress@mbari.org)
 *      Monterey Bay Aquarium Research Institute
 *      Moss Landing, CA 95039
 *    and Dale N. Chayes (dale@ldeo.columbia.edu)
 *      Lamont-Doherty Earth Observatory
 *      Palisades, NY 10964
 *
 *    See README file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * MBdatalist parses recursive datalist files and outputs the
 * complete list of data files and formats.
 * The results are dumped to stdout.
 *
 * Author:	D. W. Caress
 * Date:	October 10, 2001
 *
 * $Log: mbdatalist.c,v $
 * Revision 5.11  2007/07/05 19:16:19  caress
 * Added file copy function.
 *
 * Revision 5.10  2006/01/06 18:19:58  caress
 * Working towards 5.0.8
 *
 * Revision 5.9  2005/11/05 01:07:54  caress
 * Programs changed to register arrays through mb_register_array() rather than allocating the memory directly with mb_realloc() or mb_malloc().
 *
 * Revision 5.8  2004/10/06 19:10:52  caress
 * Release 5.0.5 update.
 *
 * Revision 5.7  2003/04/17 21:17:10  caress
 * Release 5.0.beta30
 *
 * Revision 5.6  2002/05/29 23:43:09  caress
 * Release 5.0.beta18
 *
 * Revision 5.5  2002/04/08 21:01:04  caress
 * Release 5.0.beta17
 *
 * Revision 5.4  2002/04/06 02:53:45  caress
 * Release 5.0.beta16
 *
 * Revision 5.3  2002/03/26 07:45:14  caress
 * Release 5.0.beta15
 *
 * Revision 5.2  2002/02/22 09:07:08  caress
 * Release 5.0.beta13
 *
 * Revision 5.1  2001/10/19  00:56:17  caress
 * Now tries to use relative paths.
 *
 * Revision 5.0  2001/10/12  21:07:31  caress
 * Initial revision.
 *
 *
 */

/* standard include files */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <string.h>

/* MBIO include files */
#include "../../include/mb_status.h"
#include "../../include/mb_define.h"
#include "../../include/mb_format.h"
#include "../../include/mb_status.h"
#include "../../include/mb_process.h"

static char rcs_id[] = "$Id: mbdatalist.c,v 5.11 2007/07/05 19:16:19 caress Exp $";

/*--------------------------------------------------------------------*/

int main (int argc, char **argv)
{
	char program_name[] = "mbdatalist";
	char help_message[] =  "mbdatalist parses recursive datalist files and outputs the\ncomplete list of data files and formats. \nThe results are dumped to stdout.";
	char usage_message[] = "mbdatalist [-Fformat -Ifile -N -O -P -Q -Rw/e/s/n -U -Z -V -H]";
	extern char *optarg;
	int	errflg = 0;
	int	c;
	int	help = 0;
	int	flag = 0;

	/* MBIO status variables */
	int	status = MB_SUCCESS;
	int	verbose = 0;
	int	error = MB_ERROR_NO_ERROR;

	/* MBIO read control parameters */
	char	read_file[MB_PATH_MAXLINE];
	void	*datalist;
	int	look_processed = MB_DATALIST_LOOK_UNSET;
	int	look_bounds = MB_NO;
	int	copyfiles = MB_NO;
	int	file_in_bounds = MB_NO;
	double	file_weight = 1.0;
	int	format;
	int	pings;
	int	lonflip;
	double	bounds[4];
	int	btime_i[7];
	int	etime_i[7];
	double	speedmin;
	double	timegap;
	char	fileroot[MB_PATH_MAXLINE];
	char	file[MB_PATH_MAXLINE];
	char	pwd[MB_PATH_MAXLINE];
	char	command[MB_PATH_MAXLINE];
	char	*filename;
	int	nfile = 0;
	int	make_inf = MB_NO;
	int	force_update = MB_NO;
	int	status_report = MB_NO;
	int	problem_report = MB_NO;
	int	nparproblem;
	int	ndataproblem;
	int	nparproblemtot = 0;
	int	ndataproblemtot = 0;
	int	nproblemfiles = 0;
	int	make_datalistp = MB_NO;
	
	/* output stream for basic stuff (stdout if verbose <= 1,
		output if verbose > 1) */
	FILE	*output;
	FILE	*fp;

	/* get current default values */
	status = mb_defaults(verbose,&format,&pings,&lonflip,bounds,
		btime_i,etime_i,&speedmin,&timegap);

	/* set default input to stdin */
	strcpy (read_file, "datalist.mb-1");

	/* process argument list */
	while ((c = getopt(argc, argv, "VvHhCcF:f:I:i:NnOoPpQqR:r:SsUuZz")) != -1)
	  switch (c) 
		{
		case 'C':
		case 'c':
			copyfiles = MB_YES;
			flag++;
			break;
		case 'F':
		case 'f':
			sscanf (optarg,"%d", &format);
			flag++;
			break;
		case 'H':
		case 'h':
			help++;
			break;
		case 'I':
		case 'i':
			sscanf (optarg,"%s", read_file);
			flag++;
			break;
		case 'N':
		case 'n':
			force_update = MB_YES;
			make_inf = MB_YES;
			flag++;
			break;
		case 'O':
		case 'o':
			make_inf = MB_YES;
			flag++;
			break;
		case 'P':
		case 'p':
			look_processed = MB_DATALIST_LOOK_YES;
			flag++;
			break;
		case 'Q':
		case 'q':
			problem_report = MB_YES;
			flag++;
			break;
		case 'R':
		case 'r':
			mb_get_bounds(optarg, bounds);
			look_bounds = MB_YES;
			flag++;
			break;
		case 'S':
		case 's':
			status_report = MB_YES;
			flag++;
			break;
		case 'U':
		case 'u':
			look_processed = MB_DATALIST_LOOK_NO;
			flag++;
			break;
		case 'V':
		case 'v':
			verbose++;
			break;
		case 'Z':
		case 'z':
			make_datalistp = MB_YES;
			break;
		case '?':
			errflg++;
		}

	/* set output stream */
	if (verbose <= 1)
		output = stdout;
	else
		output = stderr;

	/* if error flagged then print it and exit */
	if (errflg)
		{
		fprintf(output,"usage: %s\n", usage_message);
		fprintf(output,"\nProgram <%s> Terminated\n",
			program_name);
		error = MB_ERROR_BAD_USAGE;
		exit(error);
		}

	/* print starting message */
	if (verbose == 1 || help)
		{
		fprintf(output,"\nProgram %s\n",program_name);
		fprintf(output,"Version %s\n",rcs_id);
		fprintf(output,"MB-system Version %s\n",MB_VERSION);
		}

	/* print starting debug statements */
	if (verbose >= 2)
		{
		fprintf(output,"\ndbg2  Program <%s>\n",program_name);
		fprintf(output,"dbg2  Version %s\n",rcs_id);
		fprintf(output,"dbg2  MB-system Version %s\n",MB_VERSION);
		fprintf(output,"dbg2  Control Parameters:\n");
		fprintf(output,"dbg2       verbose:        %d\n",verbose);
		fprintf(output,"dbg2       help:           %d\n",help);
		fprintf(output,"dbg2       file:           %s\n",read_file);
		fprintf(output,"dbg2       format:         %d\n",format);
		fprintf(output,"dbg2       look_processed: %d\n",look_processed);
		fprintf(output,"dbg2       copyfiles:      %d\n",copyfiles);
		fprintf(output,"dbg2       make_inf:       %d\n",make_inf);
		fprintf(output,"dbg2       force_update:   %d\n",force_update);
		fprintf(output,"dbg2       status_report:  %d\n",status_report);
		fprintf(output,"dbg2       problem_report: %d\n",problem_report);
		fprintf(output,"dbg2       make_datalistp: %d\n",make_datalistp);
		fprintf(output,"dbg2       pings:          %d\n",pings);
		fprintf(output,"dbg2       lonflip:        %d\n",lonflip);
		fprintf(output,"dbg2       bounds[0]:      %f\n",bounds[0]);
		fprintf(output,"dbg2       bounds[1]:      %f\n",bounds[1]);
		fprintf(output,"dbg2       bounds[2]:      %f\n",bounds[2]);
		fprintf(output,"dbg2       bounds[3]:      %f\n",bounds[3]);
		fprintf(output,"dbg2       btime_i[0]:     %d\n",btime_i[0]);
		fprintf(output,"dbg2       btime_i[1]:     %d\n",btime_i[1]);
		fprintf(output,"dbg2       btime_i[2]:     %d\n",btime_i[2]);
		fprintf(output,"dbg2       btime_i[3]:     %d\n",btime_i[3]);
		fprintf(output,"dbg2       btime_i[4]:     %d\n",btime_i[4]);
		fprintf(output,"dbg2       btime_i[5]:     %d\n",btime_i[5]);
		fprintf(output,"dbg2       btime_i[6]:     %d\n",btime_i[6]);
		fprintf(output,"dbg2       etime_i[0]:     %d\n",etime_i[0]);
		fprintf(output,"dbg2       etime_i[1]:     %d\n",etime_i[1]);
		fprintf(output,"dbg2       etime_i[2]:     %d\n",etime_i[2]);
		fprintf(output,"dbg2       etime_i[3]:     %d\n",etime_i[3]);
		fprintf(output,"dbg2       etime_i[4]:     %d\n",etime_i[4]);
		fprintf(output,"dbg2       etime_i[5]:     %d\n",etime_i[5]);
		fprintf(output,"dbg2       etime_i[6]:     %d\n",etime_i[6]);
		fprintf(output,"dbg2       speedmin:       %f\n",speedmin);
		fprintf(output,"dbg2       timegap:        %f\n",timegap);
		}

	/* if help desired then print it and exit */
	if (help)
		{
		fprintf(output,"\n%s\n",help_message);
		fprintf(output,"\nusage: %s\n", usage_message);
		exit(error);
		}

	/* if make_datalistp desired then make it */
	if (make_datalistp)
		{
   		/* figure out data format and fileroot if possible */
		status = mb_get_format(verbose, read_file, 
					fileroot, &format, &error);
		sprintf(file, "%sp.mb-1", fileroot);

	    	if ((fp = fopen(file,"w")) == NULL)
		    {
		    error = MB_ERROR_OPEN_FAIL;
		    fprintf(stderr, "\nUnable to open output file %s\n", file);
		    fprintf(stderr, "Program %s aborted!\n", program_name);
		    exit(error);
		    }
		fprintf(fp, "$PROCESSED\n%s %d\n", read_file, format);
		fclose(fp);
		if (verbose > 0)
		    fprintf(output, "Convenience datalist file %s created...\n", file);
		
		/* exit unless building ancilliary files has also been requested */
		if (make_inf == MB_NO)
			exit(error);
		}

	/* get format if required */
	if (format == 0)
		mb_get_format(verbose,read_file,NULL,&format,&error);

	/* if not a datalist just output filename format and weight */
	if (format > 0)
		{
		nfile++;
		
		if (make_inf == MB_YES)
		    {
		    status = mb_make_info(verbose, force_update, 
						read_file, format, &error);
		    }
		else if (problem_report == MB_YES)
		    {
		    status = mb_pr_check(verbose, read_file, &nparproblem, &ndataproblem, &error);
		    if (nparproblem + ndataproblem > 0)
			nproblemfiles++;
		    nparproblemtot += nparproblem;
		    ndataproblemtot += ndataproblem;
			
		    }
		else
		    {
		    /* check for mbinfo file if bounds checking enabled */
		    if (look_bounds == MB_YES)
			{
			status = mb_check_info(verbose, read_file, lonflip, bounds, 
				    &file_in_bounds, &error);
			if (status == MB_FAILURE)
			    {
			    file_in_bounds = MB_YES;
			    status = MB_SUCCESS;
			    error = MB_ERROR_NO_ERROR;
			    }
			}
			
		    /* check status if desired */
		    if (status_report == MB_YES)
		    	{
			}
			
		    /* ouput file if no bounds checking or in bounds */
		    if (look_bounds == MB_NO || file_in_bounds == MB_YES)
			fprintf(output, "%s %d %f\n", read_file, format, file_weight);
		    }
		}
		
	/* else parse datalist */
	else
		{
		if ((status = mb_datalist_open(verbose,&datalist,
						read_file,look_processed,&error)) != MB_SUCCESS)
		    {
		    error = MB_ERROR_OPEN_FAIL;
		    fprintf(stderr,"\nUnable to open data list file: %s\n",
			    read_file);
		    fprintf(stderr,"\nProgram <%s> Terminated\n",
			    program_name);
		    exit(error);
		    }
		while ((status = mb_datalist_read(verbose,datalist,
					    file,&format,&file_weight,&error))
			    == MB_SUCCESS)
			{
			nfile++;
			getcwd(pwd, MB_PATH_MAXLINE);
			mb_get_relative_path(verbose, file, pwd, &error);
			if (make_inf == MB_YES)
			    {
			    status = mb_make_info(verbose, force_update,
							file, format, &error);
			    }
			else if (problem_report == MB_YES)
			    {
			    status = mb_pr_check(verbose, file, &nparproblem, &ndataproblem, &error);
		    	    if (nparproblem + ndataproblem > 0)
				nproblemfiles++;
		    	    nparproblemtot += nparproblem;
		    	    ndataproblemtot += ndataproblem;
			    }
			else if (copyfiles == MB_YES)
			    {
			    /* check for mbinfo file if bounds checking enabled */
			    if (look_bounds == MB_YES)
				{
				status = mb_check_info(verbose, file, lonflip, bounds, 
					    &file_in_bounds, &error);
				if (status == MB_FAILURE)
				    {
				    file_in_bounds = MB_YES;
				    status = MB_SUCCESS;
				    error = MB_ERROR_NO_ERROR;
				    }
				}
				
			    /* copy file if no bounds checking or in bounds */
			    if (look_bounds == MB_NO || file_in_bounds == MB_YES)
				{
				fprintf(output, "Copying %s %d %f\n", file, format, file_weight);
				sprintf(command, "cp %s* .", file);
				system(command);
				if ((filename = strrchr(file, '/')) != NULL)
					filename++;
				else
					filename = file;
				if (nfile == 1)
					system("rm datalist.mb-1");
			    	sprintf(command, "echo %s %d %f >> datalist.mb-1", 
					filename, format, file_weight);
				system(command);
			    	}
			    }
			else
			    {
			    /* check for mbinfo file if bounds checking enabled */
			    if (look_bounds == MB_YES)
				{
				status = mb_check_info(verbose, file, lonflip, bounds, 
					    &file_in_bounds, &error);
				if (status == MB_FAILURE)
				    {
				    file_in_bounds = MB_YES;
				    status = MB_SUCCESS;
				    error = MB_ERROR_NO_ERROR;
				    }
				}
			
			    /* check status if desired */
			    if (status_report == MB_YES)
		    		{
				}
				
			    /* ouput file if no bounds checking or in bounds */
			    if (look_bounds == MB_NO || file_in_bounds == MB_YES)
				fprintf(output, "%s %d %f\n", file, format, file_weight);
			    }
			    
			}
		mb_datalist_close(verbose,&datalist,&error);
		}

	/* set program status */
	status = MB_SUCCESS;
	
	/* output counts */
	if (verbose > 0)
	    {
	    fprintf(output, "\nTotal swath files:         %d\n", nfile);
	    if (problem_report == MB_YES)
	    	{
		fprintf(output, "Total files with problems: %d\n", nproblemfiles);
		fprintf(output, "Total parameter problems:  %d\n", nparproblemtot);
		fprintf(output, "Total data problems:       %d\n", ndataproblemtot);
		}
	    }

	/* check memory */
	if (verbose >= 4)
		status = mb_memory_list(verbose,&error);

	/* print output debug statements */
	if (verbose >= 2)
		{
		fprintf(output,"\ndbg2  Program <%s> completed\n",
			program_name);
		fprintf(output,"dbg2  Ending status:\n");
		fprintf(output,"dbg2       status:  %d\n",status);
		}

	/* end it all */
	exit(error);
}
/*--------------------------------------------------------------------*/
